### ISSUE TITLE
Speakers silent on Samsung Galaxy Book5 Pro 360

### ISSUE THREAD

**Issue by melvyn2** [thesofproject/linux#5242]
Similar to #5006 and older models, the speakers on this model don't work.

Headphones work mostly okay, apart from audio stopping/going silent and needing to reset the device by unplugging it or disabling it temporarily. This seems to happen mostly with games under Wine or Ryujinx.
Previously, audio used to distort often but this seems to have been fixed recently (likely #5198).

For the speakers, I tried using the kernel quirk applied to some older models (seen the patch below), but this didn't result in any change.
```patch
diff --git a/sound/pci/hda/patch_realtek.c b/sound/pci/hda/patch_realtek.c
index 3bbf5fab2..8eef0da83 100644
--- a/sound/pci/hda/patch_realtek.c
+++ b/sound/pci/hda/patch_realtek.c
@@ -10706,6 +10706,7 @@ static const struct snd_pci_quirk alc269_fixup_tbl[] = {
 	SND_PCI_QUIRK(0x144d, 0xc886, "Samsung Galaxy Book3 Pro (NP964XFG)", ALC298_FIXUP_SAMSUNG_AMP_V2_4_AMPS),
 	SND_PCI_QUIRK(0x144d, 0xc1ca, "Samsung Galaxy Book3 Pro 360 (NP960QFG)", ALC298_FIXUP_SAMSUNG_AMP_V2_4_AMPS),
 	SND_PCI_QUIRK(0x144d, 0xc1cc, "Samsung Galaxy Book3 Ultra (NT960XFH)", ALC298_FIXUP_SAMSUNG_AMP_V2_4_AMPS),
+	SND_PCI_QUIRK(0x144d, 0xc1da, "Samsung Galaxy Book5 Pro 360 (NT960QHA)", ALC298_FIXUP_SAMSUNG_AMP_V2_4_AMPS),
 	SND_PCI_QUIRK(0x1458, 0xfa53, "Gigabyte BXBT-2807", ALC283_FIXUP_HEADSET_MIC),
 	SND_PCI_QUIRK(0x1462, 0xb120, "MSI Cubi MS-B120", ALC283_FIXUP_HEADSET_MIC),
 	SND_PCI_QUIRK(0x1462, 0xb171, "Cubi N 8GL (MS-B171)", ALC283_FIXUP_HEADSET_MIC),
```

**Comment by ujfalusi** [comment]
@melvyn2, did the issue got resolved over time or still exists?
If not, please follow the https://thesofproject.github.io/latest/getting_started/intel_debug/suggestions.html# 
Attach the alsa-info outpu and try to switch to Legacy HDA stack to see if the issue with the speaker still persists.

**Comment by melvyn2** [comment]
The issue still exists, even after rebooting from windows. Here's the `alsa-info`: http://alsa-project.org/db/?f=15174857e571b0e19660dc7d25a37479762023e9
Speakers still do not work with the Legacy HDA driver. I haven't been able to test the headphone issues with it yet.

**Comment by melvyn2** [comment]
Haven't been able to reproduce the headphone bug using the Legacy HDA driver.

**Comment by ujfalusi** [comment]
@melvyn2, to sum up:
- speakers still do not work
- headset playback randomly got muted _only_ if used with the SOF kernel stack (not with legacy HDA)

I see that you are using 6.14-rc3, but I don't see information on the used sof-firmware (sof-bin) version.

The odd thing I have noted is that you are saying that the headset audio get silent mostly during gaming. I assume you are using Pipewire?
During active audio the kernel stack does not do much, it is just the DMAs are moving the data and we occasionally check the position of them and report it to userspace.
Did you noticed any lags in using the laptop even w/o audio? Stuttering and things like that.
Can you check if you have the latest BIOS and if not update it?

While this is not fixing an audio goes silent (rather it get's choppy), but if you are using PW, I would try this also, while a long shot: https://github.com/thesofproject/linux/issues/5284#issuecomment-2576977535

As for the speaker: It looks like did not work either on the previous version of the 360 (book4):
https://wiki.archlinux.org/title/Laptop/Samsung

I would try the `ALC298_FIXUP_SAMSUNG_AMP_V2_2_AMPS` version of the quirk for your device also.

7e4d4b32ab95 ("ALSA: hda/realtek: Refactor and simplify Samsung Galaxy Book init") mentions this comment with great detail, which might help with the speaker (or not): https://github.com/thesofproject/linux/issues/4055#issuecomment-2323411911

The only thing that we could help with is the headset suddenly turning to be silent, but I'm not sure if that is SOF or codec issue.

For that, it might help to add [sof-dyndbg.conf.txt](https://github.com/user-attachments/files/19069269/sof-dyndbg.conf.txt) :
```
sudo cp sof-dyndbg.conf.txt /etc/modprobe.d/sof-dyndbg.conf
```
Reboot and try to reproduce the issue and when happens, capture the kernel log if we can see anything indicative.

**Comment by melvyn2** [comment]
>     * speakers still do not work
>     * headset playback randomly got muted _only_ if used with the SOF kernel stack (not with legacy HDA)

Yes

> I see that you are using 6.14-rc3, but I don't see information on the used sof-firmware (sof-bin) version.

2025.01, from nixpkgs unstable.

> The odd thing I have noted is that you are saying that the headset audio get silent mostly during gaming. I assume you are using Pipewire? During active audio the kernel stack does not do much, it is just the DMAs are moving the data and we occasionally check the position of them and report it to userspace. Did you noticed any lags in using the laptop even w/o audio? Stuttering and things like that. Can you check if you have the latest BIOS and if not update it?

I am using pipewire. I haven't seen any graphical freezes or lag beyond some hard-freezes (requiring a reset) that are probably unrelated. BIOS is up to date.

> While this is not fixing an audio goes silent (rather it get's choppy), but if you are using PW, I would try this also, while a long shot: [#5284 (comment)](https://github.com/thesofproject/linux/issues/5284#issuecomment-2576977535)
> 
> As for the speaker: It looks like did not work either on the previous version of the 360 (book4): https://wiki.archlinux.org/title/Laptop/Samsung
> I would try the `ALC298_FIXUP_SAMSUNG_AMP_V2_2_AMPS` version of the quirk for your device also.

I'll give both of these a shot and report back.

> [7e4d4b3](https://github.com/thesofproject/linux/commit/7e4d4b32ab9532bd1babcd5d0763d727ebb04be0) ("ALSA: hda/realtek: Refactor and simplify Samsung Galaxy Book init") mentions this comment with great detail, which might help with the speaker (or not): [#4055 (comment)](https://github.com/thesofproject/linux/issues/4055#issuecomment-2323411911)
> 
> The only thing that we could help with is the headset suddenly turning to be silent, but I'm not sure if that is SOF or codec issue.
> 
> For that, it might help to add [sof-dyndbg.conf.txt](https://github.com/user-attachments/files/19069269/sof-dyndbg.conf.txt) :
> 
> ```
> sudo cp sof-dyndbg.conf.txt /etc/modprobe.d/sof-dyndbg.conf
> ```
> 
> Reboot and try to reproduce the issue and when happens, capture the kernel log if we can see anything indicative.

Will also try this too.

**Comment by melvyn2** [comment]
The pipewire fix didn't work, nor did the V2_2 ALC fixup. Here's a kernel log from the headset issue:

[dmesg_audio_cut.log](https://github.com/user-attachments/files/19116536/dmesg_audio_cut.log)

**Comment by ujfalusi** [comment]
@melvyn2, there seams to be constant stream stop/prepare/hw_params/start loop going on, which implies something is triggering xrun.

Can you verify that the PW config is in use by
```
pw-dump | grep -A3 headroom
```
It should get back with `1024`, like:
```
        "api.alsa.headroom": 1024,
        "api.alsa.open.ucm": true,
        "api.alsa.path": "hw:sofhdadsp",
...
            "name": "api.alsa.headroom",
            "description": "Headroom",
            "type": { "default": 1024, "min": 0, "max": 8192 },
```

**Comment by melvyn2** [comment]
You were right, I had mis-quoted some items when I put it into my nixOS configuration. Now that it's applied, the headphone issues seem to be gone. Speakers still do not work, as expected.

**Comment by ujfalusi** [comment]
@melvyn2, can you confirm that the pipewire 'headroom' config made the headset issue to go away and now the audio do not deteriorate?
@kv2019i, FYI

**Comment by melvyn2** [comment]
Yes I am certain that it was the headroom increase that fixed it. I made sure that the SOF stack was enabled, and it was not working with the default values previously.

**Comment by ujfalusi** [comment]
@melvyn2 , thanks for the confirmation.
